doc/experimental_ui_api.md

```diff
@@ -0,0 +1,2027 @@
+          equivalent to `context:get_state()`.
+        * `param` (table): The parameter table for this call to the builder
+          function.
+        * The function should return a freshly created `ui.Window` object.
+    * `player` (string): The player the context is associated with.
+    * `state` (table, optional): The initial state table for the window. If not
+      provided, defaults to an empty table.
+
+### Methods
+
+* `open([param])`: Builds a window and shows it to the player. Does nothing if
+  the window is already open.
+    * `param` (table, optional): The parameter table for this call to the
+      builder function. If not provided, defaults to an empty table.
+    * Returns `self` for method chaining.
+* `update([param])`: Updates a window by rebuilding it and propagating the
+  changes to the player. Does nothing if the window is not open.
+    * `param` (table, optional): The parameter table for this call to the
+      builder function. If not provided, defaults to an empty table.
+    * Returns `self` for method chaining.
+* `reopen([param])`: Reopens a window by rebuilding it, closing the player's
+  old window, and showing the new window to the player atomically. Does nothing
+  if the window is not open.
+    * `param` (table, optional): The parameter table for this call to the
+      builder function. If not provided, defaults to an empty table.
+    * Returns `self` for method chaining.
+* `close()`: Closes a window that is currently shown to the player. Does
+  nothing if the window is not currently open.
+    * Returns `self` for method chaining.
+* `is_open()`: Returns true if the window is currently open, otherwise false.
+* `get_builder()`: Returns the builder function associated with the context.
+* `get_player()`: Returns the player associated with the context.
+* `get_state()`: Returns the state table associated with the context.
+* `set_state(state)`: Sets a new state table for the context, replacing the
+  existing one.
+
+`ui.Window`
+-----------
+
+Windows represent discrete self-contained UIs as described in the [Windows]
+section.
+
+### Constructor
+
+* `ui.Window(type)(props)`: Creates a new window object.
+    * `type` (string): The window type for this window. This field cannot be
+      changed by `ui.Context:update()`.
+    * `props` (table): A table containing various fields for configuring the
+      window. See the [Fields] section for a list of all accepted fields.
+
+### Fields
+
+The following fields can be provided to the `ui.Window` constructor:
+
+* `root` (`ui.Root`, required): The root element for the element tree.
+* `theme` (`ui.Style`): Specifies a style to use as the window's theme.
+  Defaults to the theme provided by `ui.get_default_theme()`.
+* `style` (`ui.Style`): Specifies a style to apply across the entire element
+  tree. Defaults to an empty style.
+* `uncloseable` (boolean): Indicates whether the user is able to close the
```

Comment:
Love that this is an option, what if there's a bug and a modder/game maker gives them no buttons to close with? How do they exit to the main menu?

Reply:
In the future, ESC should likely open the pause menu, but for now you'd have to Alt+F4.

Reply:
Shift + Esc can now be used as of #15971.